### PR TITLE
[CSH-64] Create search service

### PR DESCRIPTION
**Pull request by Ali-Mones**
- Add search service
- Add tests passing with 100% code coverage
- Add base of course repository

### CODE REVIEW

**Review by MarwanElsays — CHANGES_REQUESTED** [review]
Use the **findbyDescriptionContaining** of the course repositoryy to make the code simpler and more readible

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package com.centralstudenthub.service;
+
+import com.centralstudenthub.entity.Course;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import com.centralstudenthub.repository.CoursesRepository;
+
+import java.util.ArrayList;
```
Comment (MuhammedAdelTaha) [inline Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:8]: Delete this unnecessary import.

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.List;
+
+@Service
+public class SearchService {
+
+    @Autowired
```
Comment (MuhammedAdelTaha) [inline Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: Use Constructor Autowired.
```
private final CoursesRepository coursesRepository;
@Autowired
public SearchService(CoursesRepository coursesRepository) {
  this.coursesRepository = coursesRepository;
}
```

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    void mockCourses() {
+        Mockito.when(coursesRepository.findAll()).thenReturn(allCourses);
+    }
+
+    @Test
+    void testLevenshteinDistance1() {
+        String s1 = "Hello!";
+        String s2 = "Hello";
+
+        int result = searchService.levenshteinDistance(s1, s2);
```
Comment (MuhammedAdelTaha) [inline Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: Inline this.

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        assertEquals(result, 1);
+    }
+
+    @Test
+    void testLevenshteinDistance2() {
+        String s1 = "Hello!";
+        String s2 = "Hllo";
+
+        int result = searchService.levenshteinDistance(s1, s2);
```
Comment (MuhammedAdelTaha) [inline Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:79]: Same here.

**Review by MuhammedAdelTaha — CHANGES_REQUESTED** [review]

File: Springboot/src/main/java/com/centralstudenthub/entity/Course.java [hunk 5]
```diff
@@ -11,6 +15,8 @@
 @Getter
```
Comment (MuhammedAdelTaha) [inline Springboot/src/main/java/com/centralstudenthub/entity/Course.java:15]: Delete Getters and Setters. They are self-contained in @Data.

**Comment by Ali-Mones** [comment]
> Use the **findbyDescriptionContaining** of the course repositoryy to make the code simpler and more readible

This doesn't accommodate for any user error during entering the search key, as the search key must be a substring of the data in the database. However searching in all the available courses using this method will accommodate for user typo mistakes.

**Comment by MarwanElsays** [comment]
Mashy ya 3am

**Comment by Ali-Mones** [comment]
- [ ] Add test for null
- [ ] Add test for empty string
- [ ] Add test for garbage input
- [ ] Add test for incorrect search for description

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 2 above)
Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: why would this be preferrable to the approach made ?

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java (hunk 3 above)
Comment (Ali-Mones) [reply Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: I believe this approach is much clearer especially for test cases

Comment (MuhammedAdelTaha) [reply Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: ايه اللي مش واضح لو حطيته فال assert نوقطة

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 2 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: ال ide بيدي warning عليه

Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: It doesn't for me, and this way is less lines of code

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java (hunk 3 above)
Comment (Ali-Mones) [reply Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: Do u mean the result variable ?

File: Springboot/src/main/java/com/centralstudenthub/entity/Course.java (hunk 5 above)
Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/entity/Course.java:15]: Done!

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java (hunk 3 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: yup.

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 1 above)
Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:8]: Done!

File: Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java (hunk 3 above)
Comment (Ali-Mones) [reply Springboot/src/test/java/com/centralstudenthub/service/SearchServiceTest.java:70]: Ah yes! Done!

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 2 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: But it is less extendable if you want to create more classes in the future.

**Review by MuhammedAdelTaha — APPROVED** [review]

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 2 above)
Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:14]: It's the same, you just add the @Autowired annotation above each defined attribute

**Comment by MarwanElsays** [comment]
change everything

**Review by mostafa336 — APPROVED** [review]

File: Springboot/src/main/java/com/centralstudenthub/service/SearchService.java (hunk 2 above)
Comment (MarwanElsays) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:13]: I agree with you

Comment (MarwanElsays) [reply Springboot/src/main/java/com/centralstudenthub/service/SearchService.java:13]: > It's the same, you just add the @Autowired annotation above each defined attribute

I Agree with you

**Review by MarwanElsays — APPROVED** [review]
